test/test_lib.py

```diff
@@ -0,0 +1,442 @@
+
+    Returns:
+        a :class:`Path` to the ``__init__.py`` of the created library.
+        This file will not have been created yet.
+    """
+    pkg = Path(topdir) / pkgname
+    try:
+        pkg.mkdir()
+    except FileExistsError:
+        pass
+    else:
+        (pkg / '__init__.py').write_text('')
+
+    lib = pkg / 'opslib' / libname
+    lib.mkdir(parents=True)
+
+    return lib / '__init__.py'
+
+
+def _flatten(specgen):
+    return [os.path.dirname(spec.origin) for spec in specgen]
+
+
+class TestLibFinder(TestCase):
+    def _mkdtemp(self) -> str:
+        tmpdir = mkdtemp()
+        self.addCleanup(rmtree, tmpdir)
+        return tmpdir
+
+    def test_single(self):
+        tmpdir = self._mkdtemp()
+
+        self.assertEqual(list(ops.lib._find_all_specs([tmpdir])), [])
+
+        _mklib(tmpdir, "foo", "bar").write_text("")
+
+        self.assertEqual(
+            _flatten(ops.lib._find_all_specs([tmpdir])),
+            [tmpdir + '/foo/opslib/bar'])
+
+    def test_multi(self):
+        tmpdirA = self._mkdtemp()
+        tmpdirB = self._mkdtemp()
+
+        dirs = [tmpdirA, tmpdirB]
+
+        expected = []
+        for top in [tmpdirA, tmpdirB]:
+            for pkg in ["bar", "baz"]:
+                for lib in ["quux", "meep"]:
+                    _mklib(top, pkg, lib).write_text("")
+                    expected.append(os.path.join(top, pkg, "opslib", lib))
+
+        self.assertEqual(
+            sorted(_flatten(ops.lib._find_all_specs(dirs))),
+            sorted(expected))
+
+    def test_cwd(self):
+        tmpcwd = self._mkdtemp()
+        os.chdir(tmpcwd)
+        self.addCleanup(os.chdir, tmpcwd)
+
+        dirs = [""]
+
+        self.assertEqual(list(ops.lib._find_all_specs(dirs)), [])
+
+        _mklib(tmpcwd, "foo", "bar").write_text("")
+
+        self.assertEqual(
+            _flatten(ops.lib._find_all_specs(dirs)),
+            ['./foo/opslib/bar'])
+
+    def test_bogus_topdir(self):
+        tmpdir = self._mkdtemp()
+
+        dirs = [tmpdir, "/bogus"]
+
+        self.assertEqual(list(ops.lib._find_all_specs(dirs)), [])
+
+        _mklib(tmpdir, "foo", "bar").write_text("")
+
+        self.assertEqual(
+            _flatten(ops.lib._find_all_specs(dirs)),
+            [tmpdir + '/foo/opslib/bar'])
+
+    def test_bogus_opsdir(self):
+        tmpdir = self._mkdtemp()
+
+        self.assertEqual(list(ops.lib._find_all_specs([tmpdir])), [])
+
+        _mklib(tmpdir, "foo", "bar").write_text('')
+
+        path = Path(tmpdir) / 'baz'
+        path.mkdir()
+        (path / 'opslib').write_text('')
+
+        self.assertEqual(
+            _flatten(ops.lib._find_all_specs([tmpdir])),
+            [tmpdir + '/foo/opslib/bar'])
+
+    def test_namespace(self):
+        tmpdir = self._mkdtemp()
+
+        self.assertEqual(list(ops.lib._find_all_specs([tmpdir])), [])
+
+        _mklib(tmpdir, "foo", "bar")  # no __init__.py  =>  a namespace package
```

Comment:
I'm not sure I understood your question, but let me try to answer it.
If you have a python package `something` that contains `opslib/otherthing`, you make `something` findable via your python path and this will let you lib.use `otherthing`.

The test on which you asked this question is for a directory that doesn't have an `__init__.py`, which means it's not something you can use directly -- you want to use things from inside it, at most. As we don't currently let you do e.g. `use("otherthing.tools")`, having `something/opslib/otherthing/tools.py` but not `something/opslib/otherthing/__init__.py` doesn't make sense.

I'm not sure how this relates to venvs and setup.py's, nor why they need to 'generate this path'. Maybe I'm missing something.

Reply:
The venvs, etc, isn't specific to this test, but specific to this PR.
What does it actually look like/feel like to use ops.lib.use in practice.
How do I write a component, what does that package look like.
I don't personally like the idea of "just use someone else's charm", because embedding the full implementation of an application in order to talk to that applications API doesn't feel correct. This seems to push us down a route of:

```
lib_foo_component/
  __init__.py (?)
  setup.py (?)
  opslib/
    __init__.py (?)
      foo/
        __init__.py (Contains LIBNAME=...)
        foo.py
        tests/
          test_foo.py
```

And then from python code you use
```
foo = ops.lib.use('foo', 2, 'john@arbash-meinel.com')
```

How do I get to `foo.FooClient`? Does it have to be imported by the `__init__.py` since I can't do "from foo import foo" or "from foo.foo import FooClient"
I don't think we want to put FooClient into `__init__.py` directly, because that makes writing the tests/test_foo.py much harder.

Reply:
as currently coded, where opslib always needs to be under something else (note one future change we discussed was also supporting it at the toplevel), and assuming this particular case is where you have an interface separate from a charm (and this is something that is not clear will be the norm), you would have something like,

```
some_client/
├── README.md
├── requirements.txt
├── setup.py
├── some
│   └── opslib
│       ├── devel
│       │   └── __init__.py
│       └── stable
│           └── __init__.py
└── tests
    └── test_some.py
```

and, yes, `ops.lib.use` only exposes the toplevel __init__, so everything that's exposed by a component needs to be there. I don't think it interferes with testing because that's better done at the toplevel, not under the source tree AFAIK.